Feature/haskell testing

OZoneGuy opened the pull request:
## Overview
<!-- Give a brief overview of your changes. -->
Added support for Haskell testing.

Addresses #6.

Needed to create a new PR due a branch rename.

This change is a
- [ ] Bug fix
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that will change existing functionality)
- [ ] Documentation update

## Description
<!-- Describe your changes in detail -->
<!-- Enumerate all
 - Areas affected
 - Features added -->
- Added support for Haskell
- Created Haskell Dockerfile
- Created `haskell_entry.sh`
- Using stack to install required package and compile program
- Improved `writeTestCaseToSuiteFile` to allow for different languages

## Motivation/Links
<!-- Why was this change needed? -->
<!-- Add any relevant links here, issues, cards or other pull requests. -->
### Notes and problems:
-  Currently, the docker container will fail due to bad volume option. It passes a relative path, while docker does not accept relative paths and requires an absolute path. Unsure how could this be fixed since I don't know how will the file system look in production.
- Needed to remove all `ulimit` commands from the entry script as it was breaking stack. It said it needed at least 70 MiB to run. And there was a problem with the `-u` command.


## How was this tested?
<!-- How were these changes tested? -->
I ran the tests using the REST API, and it behaved as expected.

## Todos
- [x] Ensure unit tests pass
- [ ] Update documentation for changes (if necessary)
- [x] Delete stale branch after merge

Commit:
feat - Wrote Dockerfile for Haskell

Wrote the docker file used to create the image for Haskell testing

Commit:
feat - Wrote Dockerfile for Haskell

Wrote the docker file used to create the image for Haskell testing

Commit:
feat - Wrote `haskell_entry.sh`

The entry script for the Haskell testing container

Commit:
feat - Wrote `haskell_entry.sh`

The entry script for the Haskell testing container

Commit:
feat - Added `haskell.dockerfile` and `haskell.entry` values

Added values in `application-local.yml`
Added variables in `DockerService.java`

Note: Most of the changes in `DockerService.java` is from the fromatter

Commit:
feat - Added `haskell.dockerfile` and `haskell.entry` values

Added values in `application-local.yml`
Added variables in `DockerService.java`

Note: Most of the changes in `DockerService.java` is from the fromatter

Commit:
feat - Added Haskell support on `CodeService.java`

Commit:
feat - Added Haskell support on `CodeService.java`

Commit:
fix - Fixed a missing brace at the end of the file

Commit:
fix - Fixed a missing brace at the end of the file

Commit:
fix - Removed resource limitations on `haskell_entry.sh`

Commented out all `ulimit` commands in `haskell_entry.sh`

There were resource problems when running the container.
1. `-u` option was problematic
2. The command said it requires at least 70MiB to run

More investigation is required to find proper resource limits.

Commit:
fix - Removed resource limitations on `haskell_entry.sh`

Commented out all `ulimit` commands in `haskell_entry.sh`

There were resource problems when running the container.
1. `-u` option was problematic
2. The command said it requires at least 70MiB to run

More investigation is required to find proper resource limits.

Commit:
fix - Improved image creation efficiency

Made the stack install a common command so it won't need to be rerun for every
image
Only `Add . .` and `ENTRYPOINT ...` are the "unique" layers

Commit:
fix - Improved image creation efficiency

Made the stack install a common command so it won't need to be rerun for every
image
Only `Add . .` and `ENTRYPOINT ...` are the "unique" layers

Commit:
fix - Fixed timeout issue for new docker images

Fixed timeout issue for new layers that take a long time to build

Increased `docker.build.timeout` to 300s from 10s

Commit:
fix - Fixed timeout issue for new docker images

Fixed timeout issue for new layers that take a long time to build

Increased `docker.build.timeout` to 300s from 10s

Commit:
fix - Added missing `break` statement

Commit:
fix - Added missing `break` statement

Commit:
feat - Improved function creation for Haskell

Commit:
feat - Improved function creation for Haskell

Commit:
feat - Made `writeTestCaseToSuiteFile` work with all Languages

Java still need to be implemented

Commit:
feat - Made `writeTestCaseToSuiteFile` work with all Languages

Java still need to be implemented

Commit:
feat - Fixed bad const for `UnsupportedProgrammingLanguageException`

Added missing argument to `String.format` call

Commit:
feat - Fixed bad const for `UnsupportedProgrammingLanguageException`

Added missing argument to `String.format` call

Commit:
feat - Now allowing the suite file to have any name

Allow the user to set the suite file name to any name, as long as it is a valid
module name

Added `-main-is` option to `stack ghc` command call

Commit:
feat - Now allowing the suite file to have any name

Allow the user to set the suite file name to any name, as long as it is a valid
module name

Added `-main-is` option to `stack ghc` command call

Commit:
feat - Automatically import HUnit to haskell suite file

Automatically add `import HUnit` to generated suite file

It is needed for all Haskell files, so best to automate it

Commit:
feat - Automatically import HUnit to haskell suite file

Automatically add `import HUnit` to generated suite file

It is needed for all Haskell files, so best to automate it

Commit:
docs - Updated README.md for Haskell

Commit:
docs - Updated README.md for Haskell

Commit:
Merge branch 'feature/haskell-testing' of github.com:puffproject/test-runner into feature/haskell-testing

OZoneGuy committed:
Fixed bad link under Haskell heading

OZoneGuy committed:
Changed link under Haskell

Changed the `Python 3` link to direct to `Sample Walkthrough` section instead of `python` section

benjaminkostiuk commented on README.md:
```diff
@@ -286,15 +286,50 @@ curl --header "Content-Type: application/json" \
+
+The differences are:
+
+1. The Suite file to be passed must import the main program file.
```
Not sure what you mean by 'passed' here...

OZoneGuy replied:
I mean the uploaded test suite file.

Something like this.
```suggestion
1. The Suite file to be uploaded must import the main program file.
```

benjaminkostiuk replied:
I see. Isin't it obvious that you need to import the thing you want to test?

OZoneGuy replied:
I thought you absolutely don't trust the user...? 😆 

I would keep it just to make it explicit.

I assume you don't mean to make it automatic? If so, I think it would be best to let the user do it, since they can import exactly what they need if there are multiple files.

benjaminkostiuk replied:
I mean their test is gonna flop otherwise 🤣, sure we can leave it in.

benjaminkostiuk commented on README.md:
````diff
@@ -286,15 +286,50 @@ curl --header "Content-Type: application/json" \
+1. The Suite file to be passed must import the main program file.
+2. A `main` function is automatically created, so the uploaded file **must not** contain one.
+3. The file automatically fully imports and exposes `HUnit` package, so avoid any conflicting symbols and names.
+4. The generated file should like something like this:
+
+```haskell
+{-
+The contents for the uploaded suite file
+This must include the module declaration.
+Must import the program file, the main submission.
+-}
+
+import HUnit
+
+main :: IO ()
+main = do
+  executedTest <- runTestTT test_{func_name}
+  print executedTest
+
+
+test_{func_name} :: Test
+test_{func_name} = {func_body}
+```
````
Can you also include what a sample Suite file should look like.

benjaminkostiuk commented on src/main/java/com/unityTest/testrunner/exception/code/UnsupportedProgrammingLanguageException.java:
```diff
@@ -4,6 +4,6 @@
 public class UnsupportedProgrammingLanguageException extends RuntimeException {
 	public UnsupportedProgrammingLanguageException(PLanguage pLanguage) {
-		super(String.format("programming language %s is not supported."));
+		super(String.format("programming language %s is not supported.", pLanguage));
```
Thanks for catching this!

benjaminkostiuk commented on src/main/java/com/unityTest/testrunner/service/CodeService.java:
```diff
@@ -172,7 +175,20 @@ private void writeTestCaseToSuiteFile(PLanguage lang, File testFile, SuiteFile s
 		bos.write(suiteFile.getContent());
 		// Write test case content to file depending on language
 		// TODO FIX THIS
```
Can remove `TODO` here

benjaminkostiuk commented on src/main/java/com/unityTest/testrunner/service/CodeService.java:
```diff
@@ -183,4 +199,26 @@ private String buildPythonTestFunction(String funcName, String body) {
 		return def.concat(Utils.indent(body, 1).concat("\n"));
 	}
+
+	private String buildHUnitTestFunction(String funcName, String body) {
+		StringBuilder resultBuilder = new StringBuilder();
+
+		// Importing Hunit
+		resultBuilder.append("import HUnit");
+
+		// Main function
+		resultBuilder.append("\n\nmain :: IO ()\n");
+		resultBuilder.append("main = do\n");
+		resultBuilder.append("  executedTest <- runTestTT test_");
+		resultBuilder.append(funcName);
+		resultBuilder.append("\n");
+		resultBuilder.append("  print executedTest\n\n");
+
+		// Test function
+		resultBuilder.append("test_");
+		resultBuilder.append(funcName);
+		resultBuilder.append(" = ");
+		resultBuilder.append(body);
+		return resultBuilder.toString();
+	}
```
This method is too verbose to be called by the `buildTestCaseCode` method. In `buildTestCaseCode` all we want to do it build the bare bones function for testing. You're building what will be displayed when people are looking through test cases written by others. They won't want to see the Hunit import etc. 
e.g. for Python all I have is
```python
def test_funcName(self):
    assert something()
```
I think it should be fine just to do the last bit + the function definition which I'm not seeing here.
```
		resultBuilder.append("test_");
		resultBuilder.append(funcName);
		resultBuilder.append(" = ");
		resultBuilder.append(body);
		return resultBuilder.toString();
```
This should produce a string like what you have in your readme
```
test_{func_name} :: Test
test_{func_name} = {func_body}
```
The rest should go in `writeTestCaseToSuiteFile`, that's where we can do the imports, main function etc.

benjaminkostiuk commented on src/main/java/com/unityTest/testrunner/service/CodeService.java:
```diff
@@ -183,4 +199,26 @@ private String buildPythonTestFunction(String funcName, String body) {
+		StringBuilder resultBuilder = new StringBuilder();
+
+		// Importing Hunit
+		resultBuilder.append("import HUnit");
```
Also you can leave out the HUnit import, they can provide that in the suite file.

benjaminkostiuk commented on src/main/java/com/unityTest/testrunner/service/CodeService.java:
```diff
@@ -183,4 +199,26 @@ private String buildPythonTestFunction(String funcName, String body) {
+		resultBuilder.append("import HUnit");
+
+		// Main function
+		resultBuilder.append("\n\nmain :: IO ()\n");
```
Not crazy about the `resultBuilder.append` everywhere. Maybe just "..." + "..." might work just fine.

OZoneGuy replied:
I believe the benefit over using `+` is that `+` creates a new `String` object, so I believe this is more efficient.

benjaminkostiuk replied:
Oh I see, could you do a longer `String.format()` if possible? At least then we can condense the lines a little.

OZoneGuy replied:
I can do some `String.format()` but I would like to keep some things in different lines to show that they are in different in the final generation. Just a personal preference.

benjaminkostiuk commented on src/main/resources/application-local.yml:
```diff
@@ -78,13 +78,13 @@ runner:
   build:
     dir: /code    # Working directory to construct in docker image
-    timeout: 10    # Timeout, in seconds, of docker build process
+    timeout: 300    # Timeout, in seconds, of docker build process
```
This needs to be reset back to something lower :).

OZoneGuy replied:
Really depends on how fast the system creates the image from scratch. Unless we have premade images, then the number can go back down to 10. Perhaps add a note to change this before prod and to figure a solution to this issue.
```suggestion
    # TODO: Fix long image creation time for images being created for the first time.
    timeout: 300    # Timeout, in seconds, of docker build process

```

benjaminkostiuk replied:
Yeah, you know what for locally with people running it for the first time it's not a big deal we can leave it at 300. The prod configuration will be managed externally anyways. We should include a base Haskell file though.

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/Dockerfile:
```diff
@@ -1,12 +1,16 @@
 FROM haskell:8
```
I'll have to run some tests locally but how big are the images with this?

OZoneGuy replied:
IIRC, 200-300 MB, these are the only images of Haskell I could find. Unless we want to create custom Haskell images which don't have as much in them. Which I don't think is possible; it just adds GHC and its tools, stack, and cabal and their respective tools.

benjaminkostiuk replied:
Wow that's alot. I'll do some digging but I think you might be right. I'll have to start looking into reducing the number of images we're creating soon.

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/haskell_entry.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
+
+# # Set virtual memory limit (Should be big) (in kbytes)
+# ulimit -v $PF_VIR_MEM
+# ulimit -Hv $PF_VIR_MEM
+
```
These need to be included, otherwise programs can consume as many resources as they want.

OZoneGuy replied:
I agree. But I still need to figure out the minimum to give them. As the PR says, the stack command says it needs at least 70 MiB.

I can try playing with the numbers asap and push something.

benjaminkostiuk replied:
Looking at the conversion I think it's like 1 MiB to 1050 kbytes. So you'll need at least 73400, which is less than I have here I think.

benjaminkostiuk replied:
We may need separate environment files for each language if they're drastically different.

OZoneGuy replied:
I think we can definitely do that.

And I thought it was in bytes, not in kbytes... So, I definitely made it way bigger.

I will change that now and push it. But I can't do anything major this late.

OZoneGuy replied:
BTW, I couldn't find any documentation on the `-u` option for `ulimit`. What is it? Because it was giving me an error saying that it is not supported...

benjaminkostiuk replied:
The `-u` flag limits the number of user processes that can run and stops the program from spawning too many additional sub-processes (I think like a fork bomb). https://linuxhint.com/linux_ulimit_command/.
It's not supported on all operating systems, if it breaks locally just comment it while testing locally.

OZoneGuy replied:
It doesn't work in the container, says it is not recognized. I am not sure if this something local to me or for every Haskell container.

And I have set the `ulimit` values, check commit [f34b026](https://github.com/puffproject/test-runner/pull/20/commits/f34b026c7cdf7d3283c2086cdcad7d4dd81fa4fd)

LMK what you think so I can merge/rebase

benjaminkostiuk replied:
Works for me, if the values work for you then 👍

benjaminkostiuk commented on src/main/resources/runner/environments/reg.env:
```diff
@@ -1,4 +1,4 @@
-PF_VIR_MEM=50000
+PF_VIR_MEM=100000000
 PF_STK_MEM=8000
```
Did you try bumping up the stack memory + max memory as well as virtual memory. That might have been why Stack was complaining.

OZoneGuy replied:
No, I am not sure about the command and I need to look it up before I make changes I am confident in.

benjaminkostiuk reviewed (commented):
This is a really good start! I'll try to make time this week to test some things locally as well. 
For your local path comment, I'm using Docker for windows and it was working fine for me. Did you set the working directory in `application-local.yml`? `runner.dir` determines where the service copies files and creates the docker images.

Commit:
docs - Improved Haskell instructions

Changed wording on the first step

Commit:
chore - Removed useless `TODO` comment

Removed `TODO` comment that was already dealt with in previous commit

Commit:
feat - Made `buildHUnitTestFunction` less verbose

Moved `main` function and `import` to `writeTestCaseToSuiteFile`
Using `String.format` to combine multiple `append` into one

Commit:
docs - Added test suite example for Haskell

Commit:
fix - Fixed bad HUnit import

Commit:
feat - Re-enabled and set `ulimit` values

Removed `-u` option, doesn't work in container
Increased $PF_VIR_MEM, required for compilation
Increased $PF_MAX_FILE_SZ, required for compilation/linking

NOTE: We will probably need to increase $PF_VIR_MEM for prod, bigger programs
might require more memory to compile

Commit:
feat - Muted successful compilation message

Commit:
fix - Fixed bad entry point

Fixed wrong entry file name

Commit:
fix - Undid troubleshooting change

Changed `Test.hs` in command back to `"$TEST_FILE_NAME"`

benjaminkostiuk commented on src/main/java/com/unityTest/testrunner/service/CodeService.java:
```diff
@@ -171,8 +174,33 @@ private void writeTestCaseToSuiteFile(PLanguage lang, File testFile, SuiteFile s
+				StringBuilder haskellBuilder = new StringBuilder();
+
+				// Importing Hunit
+				haskellBuilder.append("import Test.HUnit\n\n");
```
This should prepend a `\n` incase the suite file doesn't end on an empty line.

OZoneGuy replied:
Something like this?
```suggestion
				haskellBuilder.append("\nimport Test.HUnit\n\n");

```

benjaminkostiuk replied:
Pretty much 😆

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/haskell_entry.sh:
```diff
@@ -0,0 +1,27 @@
+# do the same command for the list of given dependencies.
+
+# Build and run test file
+timeout -s TERM $PF_TIMEOUT stack ghc "$TEST_FILE_NAME" -- -main-is $(basename "$TEST_FILE_NAME" .hs).main -o test > /dev/null
```
2 things

- Is the `> /dev/null` needed here. Do you think it would be productive to return the compiling and linking steps as well as the result?
- Can you wrap this compiling command and check for a non-zero exit code. Currently when it fails the code catches a `CONSTRAINT_VIOLATION_ERROR`, I want to return a `COMPILATION_ERROR`. If this returns a non-timeout, non-zero exit code then throw a custom unused error code and catch it in `DockerService` to return `COMPILATION_ERROR`. (In the switch line 182 `DockerService` add a new case)

OZoneGuy replied:
For the `> /dev/null`; sure, we can pass the message.

Does it also return `CONSTRAINT_VIOLATION_ERROR` on actual compilation fail? Or only when the `ulimit` is broken, or it takes too long? I can try to test to see if this is the case. And I can definitely wrap it and return a custom error code if needs be.

benjaminkostiuk replied:
When I broke the syntax in the source test file that's what it gave me. Ulimit returns different codes based on what is broken but you can to see what you get.

OZoneGuy replied:
Addressed in recent commits

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/haskell_entry.sh:
```diff
@@ -0,0 +1,27 @@
+
+# Build and run test file
+timeout -s TERM $PF_TIMEOUT stack ghc "$TEST_FILE_NAME" -- -main-is $(basename "$TEST_FILE_NAME" .hs).main -o test > /dev/null
+timeout -s TERM $PF_TIMEOUT ./test
```
`DockerService` checks for exit codes to determine pass or fail, 0 = `PASS`, 1 = `FAIL` + other error codes. Even if the test fails HUnit is returning a 0 exit code so it's returing a `PASS`. Can you make so that if the test fails the command returns a 1 exit code? Maybe use Haskell library [`System.Exit`](https://hackage.haskell.org/package/base-4.15.0.0/docs/System-Exit.html)?

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/Dockerfile:
```diff
@@ -1,12 +1,16 @@
 
-ADD . .
-RUN chmod a+x haskell_entry.sh
+RUN chown -R $USER /code
```
Can we swap `/code` for `$DIR`?
```suggestion
RUN chown -R $USER $DIR
```

benjaminkostiuk commented on README.md:
```diff
@@ -286,15 +286,58 @@ curl --header "Content-Type: application/json" \
+Must import the program file, the main submission.
+-}
+
+import HUnit
```
This should read `Test.HUnit` to match what you're doing in code.

OZoneGuy replied:
Yeah, that was a mistake on my part. It should be `Test.HUnit`

```suggestion
import Test.HUnit
```

benjaminkostiuk commented on README.md:
````diff
@@ -286,15 +286,58 @@ curl --header "Content-Type: application/json" \
 }
 ```
 
+### Haskell
````
Can you move the Haskell section under the Python section instead of after the walkthrough?

OZoneGuy replied:
Addressed in [643e4e8](https://github.com/puffproject/test-runner/pull/20/commits/643e4e8304b9c8f56e37ac4289373612e888a70c)

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/Dockerfile:
```diff
@@ -1,12 +1,16 @@
-# TODO ADD ANY OTHER DEPENDENCIES
 
 ARG DIR
+ARG USER=jeff # Consider making this an env variable
```
Just to stay consistent
```suggestion
ARG USER=appuser
```

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/Dockerfile:
```diff
@@ -1,12 +1,16 @@
+
+ADD . .
+# This build the project and run the tests
+ENTRYPOINT ["sh", "./haskell_entry.sh"]
```
Since we have bash might as well use it.
```suggestion
ENTRYPOINT ["/bin/bash", "./haskell_entry.sh"]
```

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/haskell_entry.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
```
```suggestion
#!/bin/bash
```

benjaminkostiuk commented on src/main/resources/runner/environments/reg.env:
```diff
@@ -1,6 +1,6 @@
-PF_VIR_MEM=50000
+PF_VIR_MEM=377850
```
I like round numbers 😄, unless there's a reason for this specific number.
```suggestion
PF_VIR_MEM=400000
```

benjaminkostiuk reviewed (changes requested):
Sorry for all the changes, I managed to run it on my machine and it seems to work 👍 so that's great. The only hiccup is gonna be image size for Haskell down the road...we might have to create our own in alpine or something cause 4 GB is too much.

Could you also bootstrap a sample project + test cases for haskell for easy testing, similar to what I did for python? I'll push what I did and you can build it out a little maybe. Look in `bootstrap_data.sql` to get started.

benjaminkostiuk committed:
Move all sample python files into python folder

benjaminkostiuk committed:
Add base sample files for Haskell run

OZoneGuy committed:
Rounding the `PF_VIR_MEM` number

Co-authored-by: Ben Kostiuk <kostiukb@mcmaster.ca>

OZoneGuy committed:
Using bash instead of `sh`

Co-authored-by: Ben Kostiuk <kostiukb@mcmaster.ca>

OZoneGuy committed:
Using bash instead of `sh`

Co-authored-by: Ben Kostiuk <kostiukb@mcmaster.ca>

OZoneGuy committed:
Keeping the user name consistent amongst Docker images

Co-authored-by: Ben Kostiuk <kostiukb@mcmaster.ca>

OZoneGuy committed:
Keeping the user name consistent amongst Docker images

Co-authored-by: Ben Kostiuk <kostiukb@mcmaster.ca>

benjaminkostiuk committed:
Update README.md with Test.HUnit typo

Co-authored-by: Omar Alkersh <oalkersh@protonmail.com>

benjaminkostiuk committed:
Prepend newline to HUnit import in test suite file building

Co-authored-by: Omar Alkersh <oalkersh@protonmail.com>

benjaminkostiuk commented on src/main/resources/runner/docker/haskell/Dockerfile:
```diff
@@ -1,12 +1,16 @@
+USER $USER
 
-ENTRYPOINT ["sh", "haskell_entry.sh"]
+RUN stack install HUnit
```
Can we move this instruction up to after `FROM haskell:8` since it's so large to optimize layers?

OZoneGuy replied:
I am assuming you are talking about the `RUN stack...` instruction. I say it is best to leave it after the `USER $USER` instruction since it installs the package locally, for the user. So calling it earlier will install `HUnit` for root and will need to be reinstalled for `$USER`.

It comes before `ADD . .`, so it shouldn't need to be rerun for every test case. So, the store image will not be small, but every run will take a significantly short time to build the image. If I understood you correctly.

benjaminkostiuk replied:
Yeah I meant' the `stack install` instruction. I didn't know we would need to re-install it for the user if it was already installed for root. Don't worry about moving it then.

Commit:
feat - Now return exit code 1 on test failure and code 2 on error

Program now exits with code 1 if any test case failures and exit code 2 if there
are any runtime errors

Checking that number of errors and number of failures are 0

Before program used to exit with code 0 regardless of test results, now returns
the correct exit code on test failure

Commit:
feat - Now handling compilation errors for Haskell

Return code 3 on compilation error and code 137, `TIMEOUT_ERROR`, on command timeout

Before it used to return runtime error or constraint error on compilation errors

Commit:
docs - Updated and relocated Haskell documentation

Moved Haskell section under python and above Walkthrough
Updated the sample generated test file

OZoneGuy committed:
Merge branch 'master' into feature/haskell-testing

OZoneGuy commented:
I am assuming everything is good now? If so, could you approve the PR.

benjaminkostiuk commented on src/main/java/com/unityTest/testrunner/service/CodeService.java:
```diff
@@ -171,8 +174,54 @@ private void writeTestCaseToSuiteFile(PLanguage lang, File testFile, SuiteFile s
+				// TODO;
+				break;
+			case HASKELL:
+				// TODO: Consider putting code in file and replacing function name and body into read string
```
I think this is necessary at this point. It's not exactly practical to make changes to.

benjaminkostiuk replied:
> I am assuming everything is good now? If so, could you approve the PR.

Almost!

OZoneGuy replied:
The suite template, should I save it with the Dockerfile and the entry script? Or should I create a new directory for it? If so, lmk where do want it.

And I think we should save the path in the application.yml file under `docker.build.haskell.suite-template`. Lmk if this sounds good or where do want it before I push the changes.

benjaminkostiuk replied:
That's perfect!

benjaminkostiuk commented on README.md:
```diff
@@ -169,26 +169,80 @@ from pytest import *    # Python import REQUIRED
+      then do
+      print $ "Test ran with " ++ show (failures executedTest) ++ " failures."
+      exitWith $ ExitFailure 1
+      else print "Test raun with problems!"
```
Typo

OZoneGuy replied:
```suggestion
      else print "Test(s) ran with no problems!"
```
I think this should do it...

benjaminkostiuk committed:
Update README.md to fix typo

Co-authored-by: Omar Alkersh <oalkersh@protonmail.com>

Commit:
feat - Using a template file for Haskell suite file

Storing the Haskell suite template in a text file for later read and formatting.

Easier management and more practical changes than using StringBuilder

Commit:
docs - Indicated where to find the Haskell suite template

benjaminkostiuk reviewed (approved):
You'll need to add the new `suite-template` property in `application-test.yml` for the automated tests to pass. Otherwise looks good! Feel free to squash and merge when ready.

Don't forget to give a little time for the **Bump package version** Actions script to finish after merging before deleting the branch.

Commit:
test - Added required data to application-test.yml
